app/models/concerns/orchestration/dns.rb

```diff
@@ -8,173 +8,108 @@ module Orchestration::DNS
   end
 
-  def reverse_dns?
-    # host.managed? and managed? should always come first so that orchestration doesn't
-    # even get tested for such objects
-    (host.nil? || host.managed?) && managed? && hostname.present? && ip_available? && !subnet.nil? && subnet.dns? && SETTINGS[:unattended]
+  def dns?
+    dns_ready? && ip_available? && domain.present? && domain.proxy.present?
   end
 
-  def rebuild_dns
-    logger.info "DNS not supported for #{name}, #{ip}, skipping orchestration rebuild" unless dns?
-    logger.info "Reverse DNS not supported for #{name}, #{ip}, skipping orchestration rebuild" unless reverse_dns?
-    return true unless dns? || reverse_dns?
-    del_dns_a_record_safe
-    del_dns_ptr_record_safe
-    a_record_result, ptr_record_result = true, true
-    begin
-      a_record_result = recreate_a_record if dns?
-      ptr_record_result = recreate_ptr_record if reverse_dns?
-      a_record_result && ptr_record_result
-    rescue => e
-      Foreman::Logging.exception "Failed to rebuild DNS record for #{name}, #{ip}", e, :level => :error
-      false
-    end
+  def dns6?
+    dns_ready? && ip6_available? && domain.present? && domain.proxy.present?
   end
 
-  def dns_a_record
-    return unless dns? or @dns_a_record
-    handle_validation_errors do
-      @dns_a_record ||= Net::DNS::ARecord.new dns_record_attrs
-    end
+  def reverse_dns?
+    dns_ready? && ip_available? && subnet.present? && subnet.dns?
   end
 
-  def dns_ptr_record
-    return unless reverse_dns? or @dns_ptr_record
-    handle_validation_errors do
-      @dns_ptr_record ||= Net::DNS::PTR4Record.new reverse_dns_record_attrs
-    end
+  def reverse_dns6?
+    dns_ready? && ip6_available? && subnet6.present? && subnet6.dns?
   end
 
-  def del_dns_a_record_safe
-    if dns_a_record
-      begin
-        del_dns_a_record
-      rescue => e
-        Foreman::Logging.exception "Proxy failed to delete DNS a_record for #{name}, #{ip}", e, :level => :error
-      end
+  def rebuild_dns
+    feasible = {}
+    DnsInterface::RECORD_TYPES.each do |record_type|
+      feasible[record_type] = dns_feasible?(record_type)
+      logger.info "DNS record type #{record_type} not supported for #{name}, skipping orchestration rebuild" unless feasible[record_type]
     end
-  end
+    return true unless feasible.any?
+
+    results = {}
+
+    DnsInterface::RECORD_TYPES.each do |record_type|
+      results[record_type] = true
+      del_dns_record_safe(record_type)
 
-  def del_dns_ptr_record_safe
-    if dns_ptr_record
       begin
-        del_dns_ptr_record
+        results[record_type] = recreate_dns_record(record_type) if dns_feasible?(record_type)
       rescue => e
-        Foreman::Logging.exception "Proxy failed to delete DNS ptr_record for #{name}, #{ip}", e, :level => :error
+        Foreman::Logging.exception "Failed to rebuild DNS record for #{name}(#{ip}/#{ip6})", e, :level => :error
+        return false
```

Comment:
We do need it. The diff is a little misleading. `results.values.all?` comes after this block.